.github/actions/cache/action.yml

```diff
@@ -0,0 +1,38 @@
+name: Cache
+description: 'A wrapper around `actions/cache` which exposes an additional `read-only` input to conditionally write new cache entries'
+
+inputs:
+  path:
+    description: 'A list of files, directories, and wildcard patterns to cache and restore'
+    required: true
+  key:
+    description: 'An explicit key for restoring and saving the cache'
+    required: true
+  read-only:
+    description: 'Do not write cache artifacts in the case of a missed cache'
+    default: 'false'
+    required: false
+
+outputs:
+  cache-hit:
+    description: 'A boolean value to indicate an exact match was found for the primary key'
+    value: ${{ steps.cache.outputs.cache-hit || steps.restore.outputs.cache-hit }}
+
+runs:
+  using: composite
+  steps:
+
+    - id: restore
+      uses: actions/cache/restore@v3
+      if: ${{ inputs.read_only }}
+      with:
+        path: ${{ inputs.path }}
+        key: ${{ inputs.key }}
+
+    - id: cache
+      uses: actions/cache@v3
+      if: ${{ !inputs.read_only }}
+      with:
+        path: ${{ inputs.path }}
+        key: ${{ inputs.key }}  
```

Comment:
Instead of this shared workflow, we can always use read and then conditionally use save

Reply:
I think this is cleaner though. Otherwise we need to either:
- pass key/path to restore and save separately and do synchronisation manually.
- have key/path as global variables in a job

This also keeps us closer to the standard `actions/cache` usage.

Reply:
The upsides of using save separately is that we can cache/save deps before running tests so a cache will be created even if the tests fail (and speed up next run).

Reply:
We could also rework this action to run the save step on the condition `${{ always() && !inputs.read_only }}`.

Reply:
Ah, composite actions don't support post steps.